File: Scorpio.Conversion.Runtime/Go/ScorpioConversionRuntime/DefaultReader.go
package ScorpioConversionRuntime

import (
	"bytes"
	"encoding/binary"
	"io"
	"time"
)

type DefaultReader struct {
	Buffer *bytes.Buffer
}

func (reader *DefaultReader) ReadBool() bool {
	return reader.ReadInt8() == int8(1)
}

// ReadInt8 读取一个 int8
func (reader *DefaultReader) ReadInt8() int8 {
	var ret int8
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadUInt8 读取一个 uint8
func (reader *DefaultReader) ReadUInt8() uint8 {
	var ret uint8
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadInt16 读取一个 int16
func (reader *DefaultReader) ReadInt16() int16 {
	var ret int16
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadUInt16 读取一个 uint16
func (reader *DefaultReader) ReadUInt16() uint16 {
	var ret uint16
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadInt32 读取一个 int32
func (reader *DefaultReader) ReadInt32() int32 {
	var ret int32
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadUInt32 读取一个 uint32
func (reader *DefaultReader) ReadUInt32() uint32 {
	var ret uint32
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadInt64 读取一个 int64
func (reader *DefaultReader) ReadInt64() int64 {
	var ret int64
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadUInt64 读取一个 uint64
func (reader *DefaultReader) ReadUInt64() uint64 {
	var ret uint64
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadFloat 读取一个 float
func (reader *DefaultReader) ReadFloat() float32 {
	var ret float32
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadDouble 读取一个 double
func (reader *DefaultReader) ReadDouble() float64 {
	var ret float64
	binary.Read(reader.Buffer, binary.LittleEndian, &ret)
	return ret
}

// ReadString 读取一个 string
func (reader *DefaultReader) ReadString() string {
	var length = reader.ReadUInt16()
	var bs = make([]byte, length)
	io.ReadFull(reader.Buffer, bs)
	return string(bs)
}

// ReadDateTime 读取一个 time
func (reader *DefaultReader) ReadDateTime() time.Time {
	return time.Unix(reader.ReadInt64()/1000, 0)
}

func (reader *DefaultReader) Close() {

}

File: Scorpio.Conversion.Runtime/CPP/Scorpio.Conversion.Runtime/DefaultReader.cpp
#include "DefaultReader.h"
#include <stdio.h>
#include <string.h>
namespace Scorpio {
    namespace Conversion {
        namespace Runtime {
            DefaultReader::DefaultReader(char* buffer, bool closeStream) {
                this->buf = buffer;
                this->buffer = buffer;
                this->closeStream = closeStream;
            }
            bool DefaultReader::ReadBool() {
                return ReadInt8() == (__int8)1;
            }
            __int8 DefaultReader::ReadInt8() {
                __int8 ret = *((__int8*)buffer);
                buffer += sizeof(__int8);
                return ret;
            }
            unsigned __int8 DefaultReader::ReadUInt8() {
                unsigned __int8 ret = *((unsigned __int8*)buffer);
                buffer += sizeof(unsigned __int8);
                return ret;
            }
            __int16 DefaultReader::ReadInt16() {
                __int16 ret = *((__int16*)buffer);
                buffer += sizeof(__int16);
                return ret;
            }
            unsigned __int16 DefaultReader::ReadUInt16() {
                unsigned __int16 ret = *((unsigned __int16*)buffer);
                buffer += sizeof(unsigned __int16);
                return ret;
            }
            __int32 DefaultReader::ReadInt32() {
                __int32 ret = *((__int32*)buffer);
                buffer += sizeof(__int32);
                return ret;
            }
            unsigned __int32 DefaultReader::ReadUInt32() {
                unsigned __int32 ret = *((unsigned __int32*)buffer);
                buffer += sizeof(unsigned __int32);
                return ret;
            }
            __int64 DefaultReader::ReadInt64() {
                __int64 ret = *((__int64*)buffer);
                buffer += sizeof(__int64);
                return ret;
            }
            unsigned __int64 DefaultReader::ReadUInt64() {
                unsigned __int64 ret = *((unsigned __int64*)buffer);
                buffer += sizeof(unsigned __int64);
                return ret;
            }
            float DefaultReader::ReadFloat() {
                float ret = *((float*)buffer);
                buffer += sizeof(float);
                return ret;
            }
            double DefaultReader::ReadDouble() {
                double ret = *((double*)buffer);
                buffer += sizeof(double);
                return ret;
            }
            __int64 DefaultReader::ReadDateTime() {
                return ReadInt64();
            }
            char* DefaultReader::ReadString() {
                unsigned __int16 length = ReadUInt16();
                char* ret = new char[length];
                memcpy(ret, buffer, length);
                buffer += length;
                return ret;
            }
            char* DefaultReader::ReadBytes() {
                int length = ReadInt32();
                char* ret = new char[length];
                memcpy(ret, buffer, length);
                buffer += length;
                return ret;
            }
            void DefaultReader::Close() {
                if (this->closeStream) {
                    delete[] buf;
                }
                this->buf = nullptr;
                this->buffer = nullptr;
            }
        }
    }
}
